# Import libraries

In [1]:
import keras
from keras.models import Sequential
from keras.layers import Dense
from keras.utils import to_categorical
from keras.layers.convolutional import Conv2D # to add convolutional layers
from keras.layers.convolutional import MaxPooling2D # to add pooling layers
from keras.layers import Flatten # to flatten data for fully connected layers
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error
import numpy as np
import pandas as pd

C:\Users\Hamad\anaconda3\lib\site-packages\scipy\__init__.py:146: UserWarning: A NumPy version >=1.16.5 and <1.23.0 is required for this version of SciPy (detected version 1.24.2
  warnings.warn(f"A NumPy version >={np_minversion} and <{np_maxversion}"


# Build the neural network

- Three hidden layers, each of 10 nodes and ReLU activation function.

In [2]:
def NN():
    
    model = Sequential()
    
    model.add(Dense(units = 10, activation = 'relu'))
    model.add(Dense(units = 10, activation = 'relu'))
    model.add(Dense(units = 10, activation = 'relu'))
    model.add(Dense(units = 1, activation = 'linear'))
    
    model.compile(optimizer='adam', loss='mean_squared_error')
    
    return model

# Read the data

In [3]:
df = pd.read_csv("concrete_data.csv")
df.head()

Cement  Blast Furnace Slag  Fly Ash  Water  Superplasticizer  \
0   540.0                 0.0      0.0  162.0               2.5   
1   540.0                 0.0      0.0  162.0               2.5   
2   332.5               142.5      0.0  228.0               0.0   
3   332.5               142.5      0.0  228.0               0.0   
4   198.6               132.4      0.0  192.0               0.0   

   Coarse Aggregate  Fine Aggregate  Age  Strength  
0            1040.0           676.0   28     79.99  
1            1055.0           676.0   28     61.89  
2             932.0           594.0  270     40.27  
3             932.0           594.0  365     41.05  
4             978.4           825.5  360     44.30

# Split the data into X and Y

In [4]:
X = df.drop('Strength', axis = 1)
y = df['Strength']

print(X.shape)
print(y.shape)

(1030, 8)
(1030,)


# Normalize the data

- Use a normalized version of the data. Recall that one way to normalize the data is by subtracting the mean from the individual predictors and dividing by the standard deviation.

In [5]:
X_norm = (X - X.mean()) / X.std()
X_norm.describe()

Cement  Blast Furnace Slag       Fly Ash         Water  \
count  1.030000e+03        1.030000e+03  1.030000e+03  1.030000e+03   
mean   2.432224e-15       -8.513686e-16  3.837815e-16  1.846743e-15   
std    1.000000e+00        1.000000e+00  1.000000e+00  1.000000e+00   
min   -1.714421e+00       -8.564718e-01 -8.467326e-01 -2.798851e+00   
25%   -8.496407e-01       -8.564718e-01 -8.467326e-01 -7.805147e-01   
50%   -7.911350e-02       -6.014861e-01 -8.467326e-01  1.607513e-01   
75%    6.586406e-01        8.003558e-01  1.001791e+00  4.885554e-01   
max    2.476712e+00        3.309068e+00  2.279976e+00  3.064159e+00   

       Superplasticizer  Coarse Aggregate  Fine Aggregate           Age  
count      1.030000e+03      1.030000e+03    1.030000e+03  1.030000e+03  
mean      -9.641155e-16      6.818710e-15    1.232571e-14  3.640022e-16  
std        1.000000e+00      1.000000e+00    1.000000e+00  1.000000e+00  
min       -1.038638e+00     -2.211064e+00   -2.239829e+00 -7.070160e-01  
25%       -1.038638e+00     -5.262618e-01   -5.317114e-01 -6.120340e-01  
50%        3.269920e-02     -6.326279e-02    7.383152e-02 -2.795973e-01  
75%        6.688058e-01      7.264077e-01    6.288606e-01  1.636517e-01  
max        4.351528e+00      2.213149e+00    2.731735e+00  5.055221e+00

# 1. Split the data into training and testing

- Randomly split the data into a training and test sets by holding 30% of the data for testing. You can use the train_test_splithelper function from Scikit-learn.

In [6]:
def data_split(X, y):
    return train_test_split(X, y, test_size=0.3)

X_train, X_test, y_train, y_test = data_split(X_norm, y)

# 2. Train the model

- Train the model on the training data using 100 epochs.

In [7]:
def model_training(X_train, y_train, verbose = 0):
    model = NN()
    model.fit(X_train, y_train, epochs = 100, verbose = verbose)
    return model

model = model_training(X_train, y_train, 1)

Epoch 1/100
23/23 [==============================] - 1s 2ms/step - loss: 1555.7825
Epoch 2/100
23/23 [==============================] - 0s 1ms/step - loss: 1528.1090
Epoch 3/100
23/23 [==============================] - 0s 1ms/step - loss: 1490.7488
Epoch 4/100
23/23 [==============================] - 0s 1ms/step - loss: 1439.7148
Epoch 5/100
23/23 [==============================] - 0s 1ms/step - loss: 1363.5079
Epoch 6/100
23/23 [==============================] - 0s 2ms/step - loss: 1248.6566
Epoch 7/100
23/23 [==============================] - 0s 1ms/step - loss: 1081.8499
Epoch 8/100
23/23 [==============================] - 0s 1ms/step - loss: 867.4045
Epoch 9/100
23/23 [==============================] - 0s 1ms/step - loss: 630.2614
Epoch 10/100
23/23 [==============================] - 0s 1ms/step - loss: 431.7713
Epoch 11/100
23/23 [==============================] - 0s 2ms/step - loss: 316.1824
Epoch 12/100
23/23 [==============================] - 0s 1ms/step - loss: 268.2464
Epoch 

23/23 [==============================] - 0s 1ms/step - loss: 110.8936
Epoch 100/100
23/23 [==============================] - 0s 1ms/step - loss: 110.6040


# 3. Evaluate the model

- Evaluate the model on the test data and compute the mean squared error between the predicted concrete strength and the actual concrete strength. You can use the mean_squared_error function from Scikit-learn.

In [8]:
def prediction(model, X_test, y_test):
    y_pred = model.predict(X_test)
    return mean_squared_error(y_test, y_pred)

print(f'MSE : {prediction(model, X_test, y_test)}')

10/10 [==============================] - 0s 2ms/step
MSE : 118.78997697881506


# 4. Repeat steps 1 to 3

- Repeat steps 1 - 3, 50 times, i.e., create a list of 50 mean squared errors.

In [12]:
MSE_lst = []

for i in range (50):
    X_train, X_test, y_train, y_test = data_split(X_norm, y)
    print(f'Model {i+1} :-')
    model = model_training(X_train, y_train)
    mse = prediction(model, X_test, y_test)
    MSE_lst.append(mse)

Model 1 :-
10/10 [==============================] - 0s 817us/step
Model 2 :-
10/10 [==============================] - 0s 480us/step
Model 3 :-
10/10 [==============================] - 0s 1ms/step
Model 4 :-
10/10 [==============================] - 0s 2ms/step
Model 5 :-
10/10 [==============================] - 0s 2ms/step
Model 6 :-
10/10 [==============================] - 0s 575us/step
Model 7 :-
10/10 [==============================] - 0s 2ms/step
Model 8 :-
10/10 [==============================] - 0s 1ms/step
Model 9 :-
10/10 [==============================] - 0s 2ms/step
Model 10 :-
10/10 [==============================] - 0s 2ms/step
Model 11 :-
10/10 [==============================] - 0s 2ms/step
Model 12 :-
10/10 [==============================] - 0s 2ms/step
Model 13 :-
10/10 [==============================] - 0s 2ms/step
Model 14 :-
10/10 [==============================] - 0s 365us/step
Model 15 :-
10/10 [==============================] - 0s 1ms/step
Model 16 :-
10/10 [=======

# Report for MSE

In [13]:
dic = {
    'MSE' : []
}

models = []

for i in range(len(MSE_lst)):
    models.append(i+1)
    dic['MSE'].append(MSE_lst[i])
    
report = pd.DataFrame(dic)
report.index = models
report

MSE
1    82.279848
2    95.929709
3   124.089611
4    50.924373
5   131.988304
6    71.124328
7   103.837587
8    91.980143
9   117.347588
10  111.131644
11  134.456103
12  113.389369
13  100.930239
14  121.727234
15   84.843844
16  110.861296
17  109.520026
18  108.314842
19   72.610144
20   79.196619
21   92.157121
22  102.498256
23   94.690857
24   64.230406
25  111.010086
26  104.453808
27  116.740031
28   74.275259
29   91.110760
30  117.071041
31  109.161068
32   53.874432
33  122.824636
34   72.321237
35   86.059991
36  110.354885
37   59.445060
38  119.545795
39   95.293465
40   95.055661
41  111.799643
42   98.491930
43   58.120497
44   49.447809
45   64.061516
46   94.549154
47  138.802516
48   87.648012
49   90.431374
50   98.037290

# Report the mean and the standard deviation of the mean squared errors.

In [14]:
print(f"Average of MSE          : {np.mean(MSE_lst)}")
print(f"Standard Deviation MSE  : {np.std(MSE_lst)}")

Average of MSE          : 96.00092895352988
Standard Deviation MSE  : 22.19849724283978


# How does the mean of the mean squared errors compare to that from Step B?

The Avg. MSE has reduced from 348.84 to 96.00.
The Standard Deviation in the MSE has reduced from 84.45 to 22.20.

Which make the model is performance is much better than the model in step B.